### PR TITLE
feat: Teach InfluxQL `last`, `first`, `min` and `max` selector functions

### LINKED ISSUE #7462: InfluxQL-IOx: Add a `PlannerError` type to ensure consistent error messages
Simplify error handling in the InfluxQL planner, with the ability to distinguish exceptional / internal errors from query errors.

## Background

Currently, errors returned in the InfluxQL planner use the `DataFusion::Internal` variant for representing exceptional or unexpected situations, such as the following:

https://github.com/influxdata/influxdb_iox/blob/9f5fec42b840e86534efeac76c721a82de825eb2/iox_query_influxql/src/plan/planner.rs#L563

The `DataFusion::Internal` error is reserved for DataFusion-specific errors, and encourages the user to raise an issue against the DataFusion repository. InfluxQL internal errors are an IOx issue.

It is still useful for the planner to identify its own inconsistency errors and encourage the user to raise an IOx issue instead. 

One approach is to add a `PlannerError` enumeration with similar intent to the `DataFusionError` type. For example, a template might look like: 

```rust
enum PlannerError {
    Internal(String),
    // etc
}

impl Debug for PlannerError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        todo!()
    }
}

impl Display for PlannerError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Internal(msg) => // write a message with the msg and encourage the user to log an IOx issue
        }
    }
}

impl Error for PlannerError {}

impl PlannerError {
    /// An error that was the result of an invalid query.
    fn user(s: &str) -> DataFusionError {
        DataFusionError::Plan(s.to_owned())
    }

    /// An unexpected error whilst planning that represents a bug in IOx.
    fn internal(s: &str) -> DataFusionError {
        DataFusionError::External(Box::new(Self::Internal(s.to_owned())))
    }
}
```

### LINKED ISSUE #7538: InfluxQL-IOx: Teach IOx the `LAST` function
Implement support for the [`LAST`](https://docs.influxdata.com/influxdb/v1.8/query_language/functions/#last) selector function.

> **Note**
> 
> Projecting a single selector with additional tags and fields will return a `NotImplemented` 
> error until #7533 is completed and the planner is updated to support projecting these additional columns.

### LINKED ISSUE #7564: InfluxQL-IOx: Teach IOx the `FIRST` function
Implement support for the [`FIRST`](https://docs.influxdata.com/influxdb/v1.8/query_language/functions/#first) selector function.

> **Note**
> 
> Projecting a single selector with additional tags and fields will return a `NotImplemented` 
> error until #7533 is completed and the planner is updated to support projecting these additional columns.

### LINKED ISSUE #7565: InfluxQL-IOx: Teach IOx the `MAX` function
Implement support for the [`MAX`](https://docs.influxdata.com/influxdb/v1.8/query_language/functions/#max) selector function.

> **Note**
> 
> Projecting a single selector with additional tags and fields will return a `NotImplemented` 
> error until #7533 is completed and the planner is updated to support projecting these additional columns.

### LINKED ISSUE #7566: InfluxQL-IOx: Teach IOx the `MIN` function
Implement support for the [`MIN`](https://docs.influxdata.com/influxdb/v1.8/query_language/functions/#min) selector function.

> **Note**
> 
> Projecting a single selector with additional tags and fields will return a `NotImplemented` 
> error until #7533 is completed and the planner is updated to support projecting these additional columns.

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #7564 
Closes #7538 
Closes #7565
Closes #7566
Closes #7462

## Selector functions

This PR teaches the InfluxQL planner how to execute a `SELECT` query that contain the `FIRST`, `LAST,`, `MAX` and `MIN` selector functions. 

### Single-selector queries

When a query contains a single selector function, such as:

https://github.com/influxdata/influxdb_iox/blob/8274d584f5be93a5c677692becb7323462546a5f/influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql#L432

is called a selector query, and also include the timestamp of for `first(usage_idle)`:

https://github.com/influxdata/influxdb_iox/blob/8274d584f5be93a5c677692becb7323462546a5f/influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected#L2196-L2200

### Single-selector queries with additional tag and field projections

These queries projection a single selector function, such as `FIRST(usage_idle)` and may include addition tag or field projections, such as:

https://github.com/influxdata/influxdb_iox/blob/acd6cff6311fce635213a32712c72f7557a662d0/influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql#L461

In this case, the query should produce the row with the `FIRST(usage_idle)` and the matching values at the same timestamp for `usage_user` and `usage_system`. This is not supported due to #7533 and will return an appropriate error:

https://github.com/influxdata/influxdb_iox/blob/acd6cff6311fce635213a32712c72f7557a662d0/influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected#L2348-L2349

### Selector aggregate queries

Selectors functions, when used with `GROUP BY TIME`:

https://github.com/influxdata/influxdb_iox/blob/acd6cff6311fce635213a32712c72f7557a662d0/influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql#L444-L446

with multiple selector functions:

https://github.com/influxdata/influxdb_iox/blob/acd6cff6311fce635213a32712c72f7557a662d0/influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql#L448-L450

or other aggregate functions:

https://github.com/influxdata/influxdb_iox/blob/acd6cff6311fce635213a32712c72f7557a662d0/influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql#L452-L454

are all regular aggregate queries.

## Error handling

This PR closes #7462 by introducing an `error` module, with the goal of unifying error handling via a number of convenience functions:

https://github.com/influxdata/influxdb_iox/blob/8274d584f5be93a5c677692becb7323462546a5f/iox_query_influxql/src/plan/error.rs#L3-L16

For cases when a `std::Error` is required, such as the `map_err` function, duplicates of the functions exist in the `map` module:

https://github.com/influxdata/influxdb_iox/blob/8274d584f5be93a5c677692becb7323462546a5f/iox_query_influxql/src/plan/error.rs#L18-L20

An example of the `map` module convenience functions:

https://github.com/influxdata/influxdb_iox/blob/8274d584f5be93a5c677692becb7323462546a5f/iox_query_influxql/src/plan/planner.rs#L769

### CODE REVIEW

File: iox_query_influxql/src/plan/error.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /// An error that was the result of an invalid InfluxQL query.
+    pub(crate) fn query(s: impl Into<String>) -> DataFusionError {
+        DataFusionError::Plan(s.into())
+    }
+
+    /// An unexpected error whilst planning that represents a bug in IOx.
+    pub(crate) fn internal(s: impl Into<String>) -> DataFusionError {
+        DataFusionError::External(Box::new(PlannerError::Internal(s.into())))
```
Comment (mhilton) [inline iox_query_influxql/src/plan/error.rs:38]: "External" meaning "internal" is sufficiently confusing that I think it deserves a comment.

File: iox_query_influxql/src/plan/planner.rs [hunk 2]
```diff
@@ -509,6 +491,65 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
+            };
+
+            // Take ownership of the alias, so we don't reallocate, and temporarily place a literal
+            // `NULL` in its place.
+            let Expr::Alias(_, alias) = std::mem::replace(&mut select_exprs[time_column_index], lit(ScalarValue::Null)) else {
+                return error::internal("time column is not an alias")
+            };
+
+            // Rewrite the `time` column projection based on a series of rules in the following
+            // order. If the query
```
Comment (mhilton) [inline iox_query_influxql/src/plan/planner.rs:508]: I'm sure there should be a punctuation mark at the end of this line. I can't remember which one though, possibly a semicolon.

File: iox_query_influxql/src/plan/planner.rs [hunk 3]
```diff
@@ -509,6 +491,65 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
+            //
+            // 1. is binning by time, project the column using the `DATE_BIN` function,
+            // 2. is a single-selector query, project the `time` field of the selector aggregate,
+            // 3. project the Unix epoch (0)
```
Comment (mhilton) [inline iox_query_influxql/src/plan/planner.rs:512]: "otherwise, project the unix epoch"?

**Review by mhilton — APPROVED** [review]

File: iox_query_influxql/src/plan/error.rs (hunk 1 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/error.rs:38]: The enumerations are both documented. Documentation for the `DataFusionError::External` enumeration clarifies what its purpose is:

```rust
/// Errors originating from outside DataFusion's core codebase. 
DataFusionError:: External 
```

Our `PlannerError::Internal` models the `DataFusion::Internal` error variant:

```rust
    // DataFusions has internal invariants that we are unable to ask the compiler to check for us.
    // This error is raised when one of those invariants is not verified during execution.
    Internal(String),
```

**Pull request merged**